rgw: ignore zoneless default realm when not configured

cbodley opened the pull request:
multisite configuration was simplified with squid, but may cause configurations that were valid in reef to fail on startup. when no `rgw_realm` or `rgw_zone` options are specifically requested and we fail to load the default zone of the default realm, try to fall back to the global "default" zone/zonegroup
    
Fixes: https://tracker.ceph.com/issues/67697

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

smanjara commented on src/rgw/driver/rados/rgw_zone.cc:
```diff
@@ -1199,6 +1199,16 @@ int SiteConfig::load(const DoutPrefixProvider* dpp, optional_yield y,
+    if (r == -ENOENT) {
+      ldpp_dout(dpp, 0) << "No rgw_zone configured, and the selected realm \""
+          << realm->name << "\" does not have a default zone." << dendl;
+      if (realm_name.empty()) {
```
you probably meant not `!realm_name.empty()` ?

cbodley replied:
from the block just above:
```c++
  // try to load a realm
  realm.emplace();
  std::string realm_name = dpp->get_cct()->_conf->rgw_realm;
  if (!realm_name.empty()) {
    r = cfgstore->read_realm_by_name(dpp, y, realm_name, *realm, nullptr);
  } else {
    r = cfgstore->read_default_realm(dpp, y, *realm, nullptr);
```

when `!realm_name.empty()`, we try to load that realm by name. it's the `realm_name.empty()` case that consults the 'default realm' that this new error message refers to

cbodley commented:
@smanjara i changed my mind on the error message here, and instead decided to restore _some_ of the fallback behavior that i had removed with ConfigStore/SiteConfig refactoring in https://github.com/ceph/ceph/pull/48482

the old fallback behavior that i _didn't_ restore dealt with the case where a specific `rgw_realm` was requested, but we fail to find the requested zone in it. this is still documented in [PendingReleaseNotes](https://github.com/ceph/ceph/pull/56990) for squid

this pr deals specifically with the presence of a default realm that doesn't contain a default zone. if the user hasn't configured `rgw_realm` or `rgw_zone`, we'll skip that incomplete default realm and try to load/create the global "default" zone/zonegroup

this should prevent upgrade issues for the default/realmless deployments, in case some unrelated realm had been created

cbodley commented:
cc @adamemerson @ivancich i think we should treat this (https://tracker.ceph.com/issues/67697) as a blocker for the final squid release

smanjara commented:
> @smanjara i changed my mind on the error message here, and instead decided to restore _some_ of the fallback behavior that i had removed with ConfigStore/SiteConfig refactoring in #48482
> 
> the old fallback behavior that i _didn't_ restore dealt with the case where a specific `rgw_realm` was requested, but we fail to find the requested zone in it. this is still documented in [PendingReleaseNotes](https://github.com/ceph/ceph/pull/56990) for squid
> 
> this pr deals specifically with the presence of a default realm that doesn't contain a default zone. if the user hasn't configured `rgw_realm` or `rgw_zone`, we'll skip that incomplete default realm and try to load/create the global "default" zone/zonegroup
> 
> this should prevent upgrade issues for the default/realmless deployments, in case some unrelated realm had been created

the changes look good. just a question. if a user mistakenly creates a non-default zone (without the --default option) within a realm, will it still fallback to the default zone? it should, because the realm is not default anymore, unless the realm is set to --default explicitly?

cbodley commented:
> if a user mistakenly creates a non-default zone (without the --default option) within a realm, will it still fallback to the default zone? it should, because the realm is not default anymore, unless the realm is set to --default explicitly?

@smanjara if rgw_realm or rgw_zone is configured, we'll ignore defaults and look them up by name

else if the realm is not set as default, we'll go straight to read_or_create_default_zone()

else if the realm _is_ default, we'll try to read its default zone. if that fails with ENOENT - either because rados object `default.zone.{realm-id}` doesn't exist (realm has no default zone) or because that object points at a nonexistent zone id (a zone was set as default but deleted) - we'll again fall back to read_or_create_default_zone()

cbodley commented:
jenkins test make check

cbodley commented:
jenkins test api

cbodley commented:
jenkins test make check

cbodley commented:
passed qa in https://pulpito.ceph.com/cbodley-2024-08-26_18:42:33-rgw-wip-67697-distro-default-smithi/
